torch/distributed/_spmd/api.py

```diff
@@ -53,3 +82,257 @@ def forward(
+    Override the tracing and transformation behavior of :meth:`~torch.distributed._spmd.compile`.
+    This is useful when any part of the model is not traceable or if you prefer
+    to not trace it due to any reason. More specifically, users can implement
+    :meth:`torch.distributed._spmd.Override.replacement` to replace an original
+    submodule with the return new submodule. The new submodule contrains
+    operations that users preferred to be traced, which simply be a dummy
+    placeholder operator. After tracing, users can implement
+    :meth:`torch.distributed._spmd.Override.transform` to transform the traced
+    graph, where the dummy placeholder operator serves as an anchor to insert
+    new sub-graphs.
+    """
+
+    @abstractmethod
+    def replacement(self, orig_submodule: torch.nn.Module) -> torch.nn.Module:
+        r"""
+        Implement this method to return a new :class:`nn.Module` instance to
+        replace the ``orig_submodule`` argument in the model. This helps if
+        ``orig_submodule`` is not traceable or should not be traced.
+
+        Args:
+            orig_submodule (class:`nn.Module`): original submodule instance to replace.
+
+        Returns:
+            A new :class:`nn.Module` instance to replace the original one.
+        """
+        pass
+
+    @abstractmethod
+    def transform(
+        self, gm: fx.GraphModule, schema_map: Dict[str, Schema]
+    ) -> fx.Graph:
+        r"""
+        Given a DTensor-expanded graph and shardig schema for every node,
+        conduct additional transformation for the sub-graph from the :class:`nn.Module`
+        returned by :meth:`torch.distributed._spmd.Override.replacement` if
+        necessary.
+
+        Args:
+            gm (:class:`fx.Graph`): a DTensor-expanded graph.
+            schema_map (Dict[str, :class:`Schema`]): a dictionary maps from node
+                name to DTensor schema.
+
+        Returns:
+            The :class:`fx.Graph` after transformation.
+        """
+        pass
+
+
+def _dtensor_expand(
+    gm: fx.GraphModule,
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    named_states: Dict[str, Any],
+    params_and_buffers: Dict[str, Any],
+) -> Tuple[fx.GraphModule, Dict[str, Schema]]:
+    flat_args, _ = pytree.tree_flatten(list(args) + list(kwargs.values()))
+
+    mesh = DeviceMesh("cuda", torch.arange(dist.get_world_size()).cuda())
+    shard_schema: Schema = Schema(mesh=mesh, placements=[Shard(0)])
+    # FIXME: allow other sharding schemas
+    replicate_schema: Schema = Schema(mesh=mesh, placements=[Replicate()])
+
+    inps, schemas = [], []
+    for a in flat_args:
+        if isinstance(a, torch.Tensor):
+            inps.append(a)
+            schemas.append(shard_schema)
+        elif isinstance(a, nn.Module) or isinstance(a, torch.optim.Optimizer):
+            # nn.Module or optimizer placeholder is captured by make_fx but
+            # never used in the graph
+            inps.append(torch.empty(0))
```

Comment:
did you try an alternative way of doing expansion?  not sure if i got it right but i was thinking 
1) here we can just mutate the input list of args, ... params  (or, return new lists/dicts), where we apply DTensor wrapping to the individual items in these lists
2) then we just return from this function instead of calling _convert_to_distributed
3a) we call _dtensor_expand before  `gm = make_fx(`, so the tracing also triggers dtensor expand automatically
3b) or-- if you want it separate, you just leave _dtensor_expand after first ` gm = make_fx(` and then put another ` gm = make_fx(gm` after _dtensor_expand

The main reason i asked is to simplify code structure, and remove maintenence burden.

Reply:
Hey @wconstab, thanks for the suggestion! Yep, that's the first thing we tried when developing SPMD. But at that time `make_fx` does not support nested tensor subclass. Not sure if this is already fixed. I can give it try.

Reply:
i was asking @bdhirsh about this yesterday.  He was kind of unsure if it would work or if you would run into an error where basically the mode key for fake tracing was running before the key for the subclass.  If that's still happening, then we should be able to fix it by coming up with a solution for properly ordering the keys.